File: worker/src/main/java/com/batchhawk/worker/polling/WorkerPoller.java
package com.batchhawk.worker.polling;

import com.batchhawk.worker.client.CompleteRunRequest;
import com.batchhawk.worker.client.NextJobResponse;
import com.batchhawk.worker.client.WorkerApiClient;
import lombok.RequiredArgsConstructor;
import lombok.extern.slf4j.Slf4j;
import org.springframework.scheduling.annotation.Scheduled;
import org.springframework.stereotype.Component;

import java.util.List;
import java.util.Optional;
import java.util.concurrent.TimeUnit;

@Slf4j
@Component
@RequiredArgsConstructor
public class WorkerPoller {

    private final WorkerApiClient apiClient;

    @Scheduled(fixedDelayString = "${batchhawk.worker.poll-interval-minutes}",
               timeUnit = TimeUnit.MINUTES)
    public void poll() {
        final Optional<NextJobResponse> job = apiClient.claimNextJob();
        if (job.isEmpty()) {
            log.debug("No roasters due for refresh — sleeping until next poll");
            return;
        }

        final NextJobResponse next = job.get();
        log.info("Claimed job runId={} roasterId={} integrationType={}",
                next.runId(), next.roasterId(), next.integrationType());

        try {
            final CompleteRunRequest result = scrape(next);
            apiClient.completeRun(next.runId().toString(), result);
            log.info("Completed runId={} status={}", next.runId(), result.status());
        } catch (final Exception e) {
            log.error("Job failed runId={}", next.runId(), e);
            apiClient.completeRun(next.runId().toString(), new CompleteRunRequest(
                    "FAILED", List.of(), List.of(), e.getMessage(), null));
        }
    }

    private CompleteRunRequest scrape(final NextJobResponse job) {
        // TODO: implement AI scraping agent
        // Use job.integrationType() to select the appropriate scraping strategy:
        //   SHOPIFY  → /products.json endpoint
        //   WOO_COMMERCE → REST API
        //   SQUARE / CUSTOM / UNKNOWN → general browser/AI agent
        throw new UnsupportedOperationException("Scraping not yet implemented");
    }
}

File: worker/src/main/kotlin/com/batchhawk/worker/polling/WorkerPoller.kt
package com.batchhawk.worker.polling

import com.batchhawk.common.CompleteRunRequest
import com.batchhawk.common.NextJobResponse
import com.batchhawk.worker.client.WorkerApiClient
import com.batchhawk.worker.scraper.RoasterScraper
import org.slf4j.LoggerFactory
import org.springframework.scheduling.annotation.Scheduled
import org.springframework.stereotype.Component
import java.util.concurrent.TimeUnit

@Component
class WorkerPoller(
    private val apiClient: WorkerApiClient,
    private val scrapers: Map<String, RoasterScraper>,
) {

    private val log = LoggerFactory.getLogger(WorkerPoller::class.java)

    @Scheduled(
        fixedDelayString = "\${batchhawk.worker.poll-interval-seconds}",
        timeUnit = TimeUnit.SECONDS,
    )
    fun poll() {
        val job = apiClient.claimNextJob() ?: run {
            log.debug("No roasters due for refresh — sleeping until next poll")
            return
        }

        log.info("Claimed job runId={} roasterId={} integrationType={}", job.runId, job.roasterId, job.integrationType)

        val result = runCatching { dispatch(job) }
            .getOrElse { e ->
                log.error("Job failed runId={}", job.runId, e)
                CompleteRunRequest("FAILED", emptyList(), e.message)
            }

        apiClient.completeRun(job.runId.toString(), result)
        log.info("Completed runId={} status={}", job.runId, result.status)
    }

    private fun dispatch(job: NextJobResponse): CompleteRunRequest {
        val scraper = scrapers[job.integrationType]
            ?: return CompleteRunRequest(
                "FAILED",
                emptyList(),
                "No scraper registered for integrationType=${job.integrationType}",
            )
        return scraper.scrape(job)
    }
}
